src/in_process_server.cpp

```diff
@@ -1809,13 +1817,49 @@ class wlcs::Surface::Impl
         &frame_callback
     };
 
+    static void on_enter(void* data, wl_surface* /*wl_surface*/, wl_output* output)
+    {
+        auto const self = static_cast<Impl*>(data);
+
+        if (self->outputs.find(output) != self->outputs.end())
+        {
+            BOOST_THROW_EXCEPTION(std::runtime_error(
+                "Got wl_surface.enter(wl_output@" +
+                std::to_string(wl_proxy_get_id(reinterpret_cast<wl_proxy*>(output))) +
+                ") for an output the surface is already on"));
+        }
+
+        self->outputs.insert(output);
```

Comment:
Non-blocking suggestion: you don't need a two-phase `find`/`insert` here. The [url=https://en.cppreference.com/w/cpp/container/set/insert]return value[/url] of `insert` tells you whether an element was inserted (and so whether there was an element there before).

Reply:
Done